update roadmap

little-guy-lxr opened the pull request.

sp98 commented on ROADMAP.md:
```diff
@@ -0,0 +1,32 @@
+
+###Raw devices Implement
+
+Not only lvm, operator could orchestrate csi driver that support raw devices. we should implement csi driver.
```
If I'm not wrong, this has to implemented in the Topolvm project, right?  If yes, has there been any discussions with the topolvm team about this to see if they are ok with this implementation?

little-guy-lxr replied:
It seems that topolvm project didn't make a plan to do that , and I think that raw devices relies on automatic device discovery .  if topolvm project do that , the architecture will change dramatically. So I want to make a separate driver to do that.

jmolmo replied:
IMHO, Agree with @sp98, the operator has the management role of topolvm, this feature probably is more on the topolvm side.

little-guy-lxr commented:
@sp98 @leelavg @jmolmo @Muyan0828  Can you see what you think is worth adding to the roadmap, and we can talk about it. thanks!

sp98 commented on ROADMAP.md:
```diff
@@ -0,0 +1,32 @@
+
+###Manage volume group that user created
+
+volume group that created by user has available storage capacity, use could config operator that could manage these volume groups.
```
can you please elaborate a bit more about how we plan to manage the volume groups?

little-guy-lxr replied:
Users could config volume group that created by themself in CR (TopolvmCluster) instead of operator do that.

sp98 commented on ROADMAP.md:
```diff
@@ -0,0 +1,32 @@
+###Node and devices selector
+
+Enrich node and device selectors that makes deployment more flexible
+
```
a lot of empty lines below.  We should remove them.

jmolmo replied:
> I don't know if I made it clear to you. Please feel free to contact me if you have any further questions.

Yes for sure, it is completely clear, @little-guy-lxr . I just commented if it is worth to have included these changes in the roadmap.

sp98 commented on ROADMAP.md:
```diff
@@ -0,0 +1,32 @@
+- [Data Migration and Backup](#Data-Migration-and-Backup)
+- [Nodes and devices selector](#Node-and-devices-selector)
+
+###Raw devices Implement
```
```suggestion
### Raw devices Implement
```

sp98 commented on ROADMAP.md:
```diff
@@ -0,0 +1,32 @@
+
+Not only lvm, operator could orchestrate csi driver that support raw devices. we should implement csi driver.
+
+###Manage volume group that user created
```
```suggestion
### Manage volume group that user created
```

sp98 commented on ROADMAP.md:
```diff
@@ -0,0 +1,32 @@
+
+volume group that created by user has available storage capacity, use could config operator that could manage these volume groups.
+
+###Data Migration and Backup
```
```suggestion
### Data Migration and Backup
```

sp98 commented on ROADMAP.md:
```diff
@@ -0,0 +1,32 @@
+
+Make cronjob that do regular backup
+
+###Node and devices selector
```
```suggestion
### Node and devices selector
```

jmolmo commented:
I think that we should include in the roadmap, the changes we already commented via email:

1. Operator SDK:
   - We would like to propose using the Operator SDK framework to write the controllers.
   - Operator SDK is based on the kubebuilder pattern and we feel it will be very useful for the product in the long run.   

2. CRD Changes:
  a) Remove the topolvm version from the CRD
   -  A single topolvm image should be used by all instances of the topolvmcluster (though only a single one is supported now). It is better to provide this in the CSV.
  b) Provide more filters for selection of devices via spec.storage.devices
  c) Label selector for nodes
   - It is tedious to provide individual node IP addresses in the CRD
 
3. CSV changes
 a) Set install mode to own namespace
 - The operator currently watches all namespaces but the code only allows a single topolvmcluster CR in the same namespace as the operator.

4. Log changes.
  - Operator log level is set to Debug by default.
  - Log messages need more details. We would like to work on it and move it to a more structured format.
 
5. PSP:
  - Remove PodSecurityPolicies from the code. PSPs are deprecated in k8s 1.21

6. Resource requests and limits should not be hardcoded in the code. We should be able to specify different values for different deployment types.

7. Operator Configuration: Use a configmap or similar to specify configuration parameters for the operator? For instance, resource limits and requests, single node deployment, etc.

little-guy-lxr commented:
> I think that we should include in the roadmap, the changes we already commented via email:
> 
> 1. Operator SDK:
>    
>    * We would like to propose using the Operator SDK framework to write the controllers.
>    * Operator SDK is based on the kubebuilder pattern and we feel it will be very useful for the product in the long run.
> 2. CRD Changes:
>    a) Remove the topolvm version from the CRD
>    
>    * A single topolvm image should be used by all instances of the topolvmcluster (though only a single one is supported now). It is better to provide this in the CSV.
>      b) Provide more filters for selection of devices via spec.storage.devices
>      c) Label selector for nodes
>    * It is tedious to provide individual node IP addresses in the CRD
> 3. CSV changes
>    a) Set install mode to own namespace
> 
> * The operator currently watches all namespaces but the code only allows a single topolvmcluster CR in the same namespace as the operator.
> 
> 4. Log changes.
> 
> * Operator log level is set to Debug by default.
> * Log messages need more details. We would like to work on it and move it to a more structured format.
> 
> 5. PSP:
> 
> * Remove PodSecurityPolicies from the code. PSPs are deprecated in k8s 1.21
> 
> 6. Resource requests and limits should not be hardcoded in the code. We should be able to specify different values for different deployment types.
> 7. Operator Configuration: Use a configmap or similar to specify configuration parameters for the operator? For instance, resource limits and requests, single node deployment, etc.

see reply：

Thank you for helping to raise a lot of issues that need to be improved for topolvm-operator. I now respond to your points one by one.

Operator SDK

I also wan to to use operator sdk framework to rearch the controllers.  operator-sdk framework could bring us a lot of  benefits. such as Product iteration and bundle managment.

CRD changes

         a)  remove the topolvm version from the CRD
          
         yes a single one is supported now. That is provided in CSV cr example. 
        if you remove the field topolvm versoin from the crd，how to solve the problem that users need to upgrade topolVM image

        b) Provide more filters for selection of devices via spec.storage.devices
        c) Label selector for nodes
         
        b and c  is the roadmap of topolvm-operator, if you teams could contribute, welcome and thank you very much

CSV changes   
  a) Set install mode to own namespace

This is a problem left over from history， We introduced converters to support multiple versions of CRD（TopolvmCluster), but OLM only support AllNameSpace for the conversoin webhook.
But now , we have removed the converser, so we could set install mod to own nameapce.
    4. Log changes.

         if you teams could contribute, welcome and thank you very much
        
     5. PSP
         
          We are scheduled to remove PSP

     6. resource requests and limits should not be hardcoded in the code
 
           I agree with you.

      7 . Operator Configuration

        Now operator support it. However, only a few configurations are supported, I think we could add some other configurations as you say  resource limits and requests, single node deployment, etc.

I don't know if I made it clear to you. Please feel free to contact me if you have any further questions.

jmolmo commented:
> I don't know if I made it clear to you. Please feel free to contact me if you have any further questions.

Yes for sure, it is completely clear, @little-guy-lxr . I just commented if it is worth to have included these changes in the roadmap.
